web3/_utils/proof.py

```diff
@@ -0,0 +1,133 @@
+from eth_utils import (
+    keccak,
+)
+import rlp
+from rlp.sedes import (
+    Binary,
+    big_endian_int,
+)
+
+from web3._utils.encoding import (
+    pad_bytes,
+    pad_hex,
+)
+
+
+def storage_position(map_key_hex, position_hex):
+    '''
+    https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getstorageat
+    Returns the patricia trie key (position) of a mapping's key.
+    This helper function is used for formating arguments to eth_getStorageAt
+    and eth_getProof.
+    @param map_key_hex is the key stored in a map
+    @param position_hex is the position where the mapping variable is defined
+        in the contract
+    '''
+    key = pad_hex(map_key_hex, 256)[2:] + pad_hex(position_hex, 256)[2:]
+    return keccak(bytes.fromhex(key)).hex()
+
+
+nibble_to_number = {}
+for i, c in enumerate('0123456789abcdef'):
+    nibble_to_number[c] = i
+
+
+class _Account(rlp.Serializable):
+    fields = [
+        ('nonce', big_endian_int),
+        ('balance', big_endian_int),
+        ('storage', Binary.fixed_length(32, allow_empty=True)),
+        ('code_hash', Binary.fixed_length(32))
+    ]
+
+
+def verify_eth_getProof(proof, root):
+    '''
+    Verify that a proof returned by eth_getProof is valid for the given
+    Patricia tree state root.
+    @param proof is a proof returned by eth_getProof
+    @param root is a Patricia tree state root returned by w3.eth.getBlock('latest').stateRoot
+    '''
+    acc = _Account(proof.nonce,
+                   proof.balance,
+                   proof.storageHash,
+                   proof.codeHash)
+    expected_value = rlp.encode(acc)
+    addr = proof.address[2:]
+    trie_key = keccak(bytes.fromhex(addr)).hex()
+    account_proof = proof.accountProof
+    if not _verify(root, trie_key, account_proof,
+                   0, 0,
+                   expected_value):
+        return False
+    for storage_proof in proof.storageProof:
+        trie_key = keccak(pad_bytes(b'\x00', 32, storage_proof.key)).hex()
+        root = proof.storageHash
+        expected_value = rlp.encode(storage_proof.value)
+        if not _verify(root, trie_key, storage_proof.proof,
+                       0, 0,
+                       expected_value):
+            return False
+    return True
+
+
+def _verify(expected_root, key, proof, key_index, proof_index, expected_value):
```

Comment:
We definitely don't want to maintain another copy of a `trie` navigator here, py-trie is the canonical one. I think we can accomplish the goal of this method using `trie.HexaryTrie.get_from_proof()`.

We probably don't want to add `py-trie` as a core dependency, but adding it to the `tests` extra seems reasonable.

I think this module is only being used for testing, in which case we can change to something like `web3._utils.test_proofs`, plus a comment at the top of the file that it's only intended to be imported in testing contexts (because the py-trie dependency is only satisfied during testing).

Reply:
Thank you for reviewing, i didn't know about py-trie :-)
My idea for adding verify_eth_getProof() is that a user should trust only the state root and after verifying the proofs (with web3py) returned by getProof() he can safely use the values in the proof object. 

So it depends if you want the proof verification to be supported by web3py or not :
- if yes the proof verification should be a public API : i can refactor and use py-trie as core dependency
- if not, web3py is just used to get the getProof() result and users need to write their own code to verify it (testing for proof verification is not even needed in this case i think)

Which do you think is suitable for web3py ?

Reply:
Cool, happy to help!

My instinct is that for now web3.py just returns the proof. The docs can demonstrate the couple of lines it takes to use py-trie to get a value from the proof.

It's possible that a project would want to get the proof only to record to a database or re-communicate over the network. So I'd say let's avoid adding the py-trie dependency until it becomes obvious that it's necessary.